modules/pytket-iqm/pytket/extensions/iqm/backends/iqm.py

```diff
@@ -0,0 +1,344 @@
+        :param username: IQM username
+        :param api_key: IQM API key
+        """
+        super().__init__()
+        self._url = url
+        config = IQMConfig.from_default_config_file()
+
+        if username is None:
+            username = config.username
+        if username is None:
+            raise IqmAuthenticationError()
+        if api_key is None:
+            api_key = config.api_key
+        if api_key is None:
+            raise IqmAuthenticationError()
+
+        with open(device) as f:
+            settings = json.load(f)
+        self._client = IQMClient(
+            self._url, settings=settings, username=username, api_key=api_key
+        )
+        self._qubits = [
+            _as_node(qb) for qb in settings["subtrees"].keys() if qb.startswith("QB")
+        ]
+        self._n_qubits = len(self._qubits)
+        if arch is None:
+            arch = _DEFAULT_COUPLING
+        coupling = [(_as_node(a), _as_node(b)) for (a, b) in arch]
+        if any(qb not in self._qubits for couple in coupling for qb in couple):
+            raise ValueError("Architecture contains qubits not in device")
+        self._arch = Architecture(coupling)
+        self._backendinfo = BackendInfo(
+            type(self).__name__,
+            settings["name"],
+            __extension_version__,
+            self._arch,
+            _GATE_SET,
+        )
+
+    @property
+    def backend_info(self) -> BackendInfo:
+        return self._backendinfo
+
+    @property
+    def required_predicates(self) -> List[Predicate]:
+        return [
+            NoClassicalControlPredicate(),
+            NoFastFeedforwardPredicate(),
+            NoMidMeasurePredicate(),
+            NoSymbolsPredicate(),
+            GateSetPredicate(_GATE_SET),
+            ConnectivityPredicate(self._arch),
+        ]
+
+    def default_compilation_pass(self, optimisation_level: int = 1) -> BasePass:
+        assert optimisation_level in range(3)
+        passes = [DecomposeBoxes(), FlattenRegisters()]
+        if optimisation_level == 1:
+            passes.append(SynthesiseTket())
+        elif optimisation_level == 2:
+            passes.append(FullPeepholeOptimise())
+        passes.append(_iqm_rebase())  # to satisfy MaxTwoQubitGatesPredicate
+        passes.append(DefaultMappingPass(self._arch))
+        passes.append(DelayMeasures())
+        passes.append(_iqm_rebase())
+        passes.append(RemoveRedundancies())
+        if optimisation_level >= 1:
+            passes.append(
+                SimplifyInitial(
+                    allow_classical=False, create_all_qubits=True, xcirc=_xcirc
+                )
+            )
+        return SequencePass(passes)
+
+    @property
+    def _result_id_type(self) -> _ResultIdTuple:
+        return (bytes, str)
+
+    def process_circuits(
+        self,
+        circuits: Sequence[Circuit],
+        n_shots: Union[None, int, Sequence[Optional[int]]] = None,
```

Comment:
Are you sure this can handle n_shots = None?

Reply:
I'll add a test to make sure.